src/storage/copier/csv_reader.cpp

```diff
@@ -0,0 +1,515 @@
+      columnSizes(tableSchema->getNumProperties()) {}
+
+BaseCSVReader::~BaseCSVReader() {}
+
+void BaseCSVReader::AddValue(std::string str_val, column_id_t& column,
+    std::vector<uint64_t>& escape_positions, bool has_quotes, uint64_t buffer_idx) {
+    if (mode == ParserMode::PARSING_HEADER) {
+        return;
+    }
+    auto length = str_val.length();
+    if (length == 0 && column == 0) {
+        row_empty = true;
+    } else {
+        row_empty = false;
+    }
+    if (!return_types.empty() && column == return_types.size() && length == 0) {
+        // skip a single trailing delimiter in last column
+        return;
+    }
+    if (column >= return_types.size()) {
+        throw CopyException(StringUtils::string_format(
+            "Error in file {}, on line {}: expected {} values per row, but got more. ", filePath,
+            linenr, return_types.size()));
+    }
+
+    // insert the line number into the chunk
+    if (!escape_positions.empty()) {
+        // remove escape characters (if any)
+        std::string old_val = str_val;
+        std::string new_val = "";
+        uint64_t prev_pos = 0;
+        for (uint64_t i = 0; i < escape_positions.size(); i++) {
+            uint64_t next_pos = escape_positions[i];
+            new_val += old_val.substr(prev_pos, next_pos - prev_pos);
+            prev_pos = next_pos + 1;
+        }
+        new_val += old_val.substr(prev_pos, old_val.size() - prev_pos);
+        escape_positions.clear();
+        str_val = new_val;
+    }
+    auto type = parse_chunk.getValueVector(column)->dataType;
+    switch (type.getLogicalTypeID()) {
+    case LogicalTypeID::INT64: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++,
+            StringCastUtils::castToNum<int64_t>(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::INT32: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++,
+            StringCastUtils::castToNum<int32_t>(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::INT16: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++,
+            StringCastUtils::castToNum<int16_t>(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::FLOAT: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++,
+            StringCastUtils::castToNum<float_t>(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::DOUBLE: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++,
+            StringCastUtils::castToNum<double_t>(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::BOOL: {
+        parse_chunk.getValueVector(column)->setValue(
+            columnSizes[column]++, StringCastUtils::castToBool(str_val.c_str(), str_val.length()));
+    } break;
+    case LogicalTypeID::STRING: {
+        parse_chunk.getValueVector(column)->setValue(columnSizes[column]++, str_val);
```

Comment:
This is incorrect. I thought we discussed this before. You should refer to `StringVector::addString`.